### DISCUSSION TITLE
Get user object without population of relations

### DISCUSSION THREAD

**Discussion by alxnkt** [strapi/strapi#4913] [Legacy Topics]
**Describe the bug**
I have users in my API, that have relations with other entities. When I'm authorizing, I'm getting user object with populated relations. This object can be very big.

**Steps to reproduce the behavior**
1. Authorize via ```http://localhost:1337/auth/local```.
2. See ```response.data.user``` object.

**Expected (desired) behavior**
User object **without** populated relations.

**Code snippets**
```
const response = await this.$axios.post('/auth/local', {
   identifier: params.email,
   password: params.password
})
```
**System**
- Node.js version: 13.5.0
- NPM version: 6.13.4
- Strapi version: v3.0.0-beta.18.3
- Database: MongoDB
- Operating system: MacOS 10.15.2

**Additional context**
Is there any way to get unpopulated user object?

**Comment by theo-stocchetti** [comment]
A similar problem is describe on this topic but for  : https://github.com/strapi/strapi/issues/4878 
Personnaly, i retrieve all relation (but only 2 small object) on authentification, my problem is for "users/me"

I'm on v3.0.0-beta.18.4 version, this last update have fix several bug that I had user custom fields relationship.

Try to update.

**Comment by alxnkt** [comment]
I think, this is more graphql-related issue.
I'm on v3.0.0-beta.18.4 too, but I'm facing same problem.

**Comment by lauriejim** [comment]
Hello! You can customize the login function to not populate all relations.

[Here](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-users-permissions/controllers/Auth.js#L71) is the function and the line you will have to customize by using the [customization](https://strapi.io/documentation/3.0.0-beta.x/concepts/customization.html#plugin-extensions) concept.

So you will have to create a `./extensions/users-permissions/controllers/Auth.js` file. And in this file copy the source one, keep only the `callback` function (just this function to override).
And update the line with the following code.

```js
const user = await strapi
        .query('user', 'users-permissions')
        .findOne(query, []);
```

I give `[]` as the second param. The second param is to specify the populated data, in this case, none.

Now I have a question for @Aurelsicoko and @alexandrebodin.
In the login payload, is it a requirement to have relations?
Or do we have to login, get the JWT token and then request `/users/me` endpoint to have all user information?

**Comment by alxnkt** [comment]
Grand merci!
Works like a charm now!

**Comment by Aurelsicoko** [comment]
@alexandrebodin Do we still use the `autoPopulate` boolean to automatically populate relationships of a model by default?

The solution will be to send parameters in the request to define which fields to populate, or not. There is a similar card on Productboard to filter/select the fields to return in the response (see https://portal.productboard.com/strapi/1-public-roadmap/c/31-exclude-include-fields-from-query-request)

**Comment by alexandrebodin** [comment]
@Aurelsicoko  we do have the `autoPopulate` param. 

Adding a `populate` param requires thinking about depth, permissions / security issues with accessing any relation. We also need to think about populate limits / fields / sorting / ordering.

**Comment by harmenjanssen** [comment]
I was actually looking for the solution to exactly this problem. 
Although it works like a charm, I'm a bit reluctant to overwrite the `Auth.js` controller file. I can see it's already had quite some changes since the file was referenced in January in this thread, which I would've missed if I had overwritten the file.

It would be nice to maybe implement an event or hook system that allows developers to hook into the core and customize certain things without having to overwrite all of the core functionality and block future updates.

**Comment by lauriejim** [comment]
Hello! That is what we plan to do.
We will announce it when it will be done.

**Reply by adamkhan** [reply]
Any update on this? I just came across the problem myself and very glad to have found the solution here!
